tests/unit/test_charm.py

```diff
@@ -261,3 +273,178 @@ def test_config_updated_on_ingress_relation_joined(harness) -> None:
     }
 
     assert yaml.safe_load(harness.charm._render_conf_file()) == expected_config
+
+
+def test_provider_info_in_databag_when_ingress_then_oauth_relation(harness):
+    harness.begin()
+    harness.set_can_connect(CONTAINER_NAME, True)
+
+    setup_ingress_relation(harness, "public")
+    setup_ingress_relation(harness, "admin")
+    relation_id, _ = setup_oauth_relation(harness)
+
+    app_data = harness.get_relation_data(relation_id, harness.charm.app)
```

Comment:
You may want to separate concerns between the charm code and the lib code. Asserting content of relation data is probably preferable to test in the lib tests since it's the lib that does this. In the charm tests you may want to simply validate that the correct lib method is called.

Reply:
But you also need to be sure that the charm is passing the correct information to the library. I could mock the library methods and assert that they are called with the correct arguments, but it seems simpler to validate that the final state is correct (ie the correct information is placed in the databag).

Reply:
Exact, I think you should do the first thing you are proposing (mocking the lib and asserting you are calling it with the correct args). If you are validating for the relation data you would be duplicating tests as you'd also be testing this in the lib tests.

Reply:
Another way to see this. If the lib doesn't write the relation data correctly, you want only 1 test to crash, ideally a test in the lib, not on something that is dependent on the lib.